shared/chat/blocking/block-modal/index.tsx

```diff
@@ -0,0 +1,346 @@
+      this.props.onClose()
+    }
+  }
+
+  getBlockFor(username: string, which: BlockType) {
+    // First get a current setting from a checkbox, if user has checked anything.
+    const {newBlocks} = this.state
+    const current = newBlocks.get(username)
+    if (current && current[which] !== undefined) {
+      return current[which] || false
+    }
+    // If we don't have a checkbox, check the store for current block value.
+    return this.props.isBlocked(username, which)
+  }
+
+  setReport = (checked: boolean) => {
+    this.setState({shouldReport: checked})
+  }
+
+  setReportReason = (reason: string) => {
+    this.setState({reportReason: reason})
+  }
+
+  setBlockFor(username: string, which: BlockType, block: boolean) {
+    const {newBlocks} = this.state
+    const current = newBlocks.get(username)
+    if (current) {
+      current[which] = block
+      newBlocks.set(username, current)
+    } else {
+      newBlocks.set(username, {[which]: block})
+    }
+    // Need to make a new object so the component re-renders.
+    this.setState({newBlocks: new Map(newBlocks)})
+  }
+
+  setBlockTeam = (checked: boolean) => {
+    this.setState({blockTeam: checked})
+  }
+
+  setExtraNotes = (text: string) => {
+    this.setState({extraNotes: text})
+  }
+
+  setIncludeTranscript = (checked: boolean) => {
+    this.setState({includeTranscript: checked})
+  }
+
+  onFinish = () => {
+    if (this.state.newBlocks.size === 0 && this.state.blockTeam && !this.state.shouldReport) {
+      // Nothing to do, just close the modal.
+      this.props.onClose()
+      return
+    }
+
+    this.setState({finishClicked: true})
+    let report: ReportSettings | undefined = undefined
+    if (this.state.shouldReport) {
+      report = {
+        extraNotes: this.state.extraNotes,
+        includeTranscript: this.state.includeTranscript,
+        reason: this.state.reportReason,
+      }
+    }
+    this.props.onFinish(this.state.newBlocks, this.state.blockTeam, report)
+  }
+
+  render() {
+    const {teamname, adderUsername} = this.props
+
+    const header = {
+      leftButton: (
+        <Kb.Text onClick={this.props.onClose} type="BodyPrimaryLink">
+          Cancel
+        </Kb.Text>
+      ),
+      title: <Kb.Icon type="iconfont-block-user" sizeType="Big" color="red" />,
+    }
+
+    if (this.props.loadingWaiting) {
+      return (
+        <Kb.Modal mode="Default" header={header}>
+          <Kb.Box style={styles.loadingAnimationBox}>
+            <Kb.Animation animationType="spinner" style={styles.loadingAnimation} />
+          </Kb.Box>
+        </Kb.Modal>
+      )
+    }
+
+    return (
+      <Kb.Modal
+        mode="Default"
+        header={header}
+        footer={{
+          content: (
+            <Kb.WaitingButton
+              label="Finish"
+              onClick={this.onFinish}
+              fullWidth={true}
+              type="Danger"
+              waitingKey={null}
```

Comment:
Shouldn't this be something so it can correctly disappear?

Reply:
The entire modal should disappear when action driven by this button finishes. I'll see if I can add proper waiting keys here though